src/components/va-pagination/va-pagination.tsx

```diff
@@ -0,0 +1,306 @@
+  };
+
+  /**
+   * Add ellipsis separator
+   * @returns {JSX} ellipsis separator
+   */
+  private addEllipsis = () =>
+    <li aria-hidden="true">
+      <span class='ellipsis'>{'\u2026'}</span>
+    </li>;
+
+  /**
+   * Build page links
+   * @param index - Current page index to build 
+   * @param Options
+   * @returns {JSX} page link
+   */
+  private buildLink = (index: number, { current, total }) => {
+    let page = index + 1;
+    if (page < 1) {
+      page = 1;
+    } else if (page >= total) {
+      page = total;
+    }
+    const isCurrent = page === current;
+    const props = this.getLinkProps({
+      page,
+      current: isCurrent ? 'page' : '',
+      className: `va-pagination-link ${isCurrent ? 'va-pagination-active' : ''}`,
+    });
+    return (
+      <li class={isCurrent ? 'va-pagination-current' : null}>
+        <a {...props}>{page}</a>
+      </li>
+    );
+  };
+
+  /**
+   * Calculates and build visible pages
+   * @param Options
+   * @returns {JSX} all visible page links
+   */
+  private buildPager = ({ current, total, edges, visible }) => {
+    let index : number;
+    const pages = [];
+
+    // max length - 2 edges - 2 ellipsis
+    const display = visible - (edges * 2) - 2;
+
+    const half = display / 2;
+    const start = Math.floor(
+      current > half
+        ? Math.max(Math.min(current - half, total - display), 0)
+        : 0
+    );
+    const end = Math.floor(
+      current > half
+        ? Math.min(current + half, total)
+        : Math.min(display, total)
+    );
+
+    // build start edge links & add ellipsis
+    if (start > 0 && edges > 0) {
+      for (index = 0; index < Math.min(edges, start); index++) {
+        pages.push(this.buildLink(index, { current, total }));
+      }
+      if (edges < start && (start - edges !== 1)) {
+        pages.push(this.addEllipsis());
+      } else if (start - edges === 1) {
+        pages.push(this.buildLink(edges, { current, total }));
+      }
+    }
+
+    // build middle links
+    for (index = start; index < end; index++) {
+      pages.push(this.buildLink(index, { current, total }));
+    }
+
+    // build ellipsis & ending edge links
+    if (end < total && edges > 0) {
+      if (
+        total - edges > end &&
+        (total - edges - end !== 1)
+      ) {
+        pages.push(this.addEllipsis());
+      } else if (total - edges - end === 1) {
+        pages.push(this.buildLink(end, { current, total }));
+      }
+      for (
+        index = Math.max(total - edges, end);
+        index < total;
+        index++
+      ) {
+        pages.push(this.buildLink(index, { current, total }));
+      }
+    }
+    return pages;
+  };
+
+  render() {
+    const opts = this.checkValidation();
+    return (
+      <nav aria-label="pagination">
+        <ul>
+          {this.addPrevLink(opts)}
+          {this.buildPager(opts)}
+          {this.addNextLink(opts)}
+        </ul>
+      </nav>
+    );
+  };
```

Comment:
I'm curious to see if we could keep the JSX more pieced together rather than having to visually jump around to the right function to look for the markup and having to stitch it together in your head. What do you think of something like this:


```js

  render() {
    const { current } = this.checkValidation();
    const prevPage = current - 1 >= 1 ? current - 1 : null;
  
    return (
      <nav aria-label="pagination">
        <ul>
          <li>
            {/** Previous button **/}
            <a {...props} disabled={prevPage === null}>
              <span class="vads-u-visibility--screen-reader">Previous</span>
              <span aria-hidden="true">{'\u2039\u00a0\u00a0'}
                <span class="hide-narrow">Prev</span>
              </span>
            </a>
          </li>
          {this.buildPager(opts)}
          {this.addNextLink(opts)}
        </ul>
      </nav>
    );
  };
```
Looking at this right now I'm not actually sure if I'm sold - I do recognize that there is a balance, because a giant wall of dozens of lines of nesting JSX could also be hard to decipher. :shrug: 
  
  Also, it seems like whether or not we have a previous or next page only determines if an `<a>` gets included as a wrapper inside of the `<li>`. Do you think it would be possible to simplify some things by using a bool to set `disabled` on the anchor tag?

Reply:
I started out that way, but the `<a>` tag doesn't have a disabled state - https://css-tricks.com/how-to-disable-links/#just-dont-do-it - I mean it is possible to set the `pointer-events` to `none` to disable it, but I think it still causes problems for screen reader users.

Either way, I'm open to the idea of moving the prev & next links into the `render` function if you think that makes it more readable.